### DISCUSSION TITLE
How to access dmp1 of RSA

### DISCUSSION THREAD

**Discussion by nagu1137** [openssl/openssl#23506] [Q&A] [answered]
WE ARE Using RSA_get0_dmp1 IN OUR LEGACY CODE , we are upgrading from 1.0.2 U to 3.1.3.

not able to find correcspoind method/macro to retrive the value using EVP_PKEY_get_bn_param in 3.1.3.

please guide us on how can we access this value dmp1

we are using this for coverting RSA KEY from PEM to BLOB format

**Comment by levitte** [comment, marked as answer]
To make parameters consistent for all RSA keys including multi-prime ones, we went with the terms "FACTOR", "EXPONENT" and "COEFFICIENT", so we have these parameter names (and macros):

| Macro name                       | parameter name     |
|----------------------------------|--------------------|
| OSSL_PKEY_PARAM_RSA_FACTOR1      | "rsa-factor1"      |
| OSSL_PKEY_PARAM_RSA_FACTOR2      | "rsa-factor2"      |
| OSSL_PKEY_PARAM_RSA_FACTOR3      | "rsa-factor3"      |
| OSSL_PKEY_PARAM_RSA_FACTOR4      | "rsa-factor4"      |
| OSSL_PKEY_PARAM_RSA_FACTOR5      | "rsa-factor5"      |
| OSSL_PKEY_PARAM_RSA_FACTOR6      | "rsa-factor6"      |
| OSSL_PKEY_PARAM_RSA_FACTOR7      | "rsa-factor7"      |
| OSSL_PKEY_PARAM_RSA_FACTOR8      | "rsa-factor8"      |
| OSSL_PKEY_PARAM_RSA_FACTOR9      | "rsa-factor9"      |
| OSSL_PKEY_PARAM_RSA_FACTOR10     | "rsa-factor10"     |
| OSSL_PKEY_PARAM_RSA_EXPONENT1    | "rsa-exponent1"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT2    | "rsa-exponent2"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT3    | "rsa-exponent3"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT4    | "rsa-exponent4"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT5    | "rsa-exponent5"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT6    | "rsa-exponent6"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT7    | "rsa-exponent7"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT8    | "rsa-exponent8"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT9    | "rsa-exponent9"    |
| OSSL_PKEY_PARAM_RSA_EXPONENT10   | "rsa-exponent10"   |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT1 | "rsa-coefficient1" |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT2 | "rsa-coefficient2" |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT3 | "rsa-coefficient3" |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT4 | "rsa-coefficient4" |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT5 | "rsa-coefficient5" |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT6 | "rsa-coefficient6" |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT7 | "rsa-coefficient7" |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT8 | "rsa-coefficient8" |
| OSSL_PKEY_PARAM_RSA_COEFFICIENT9 | "rsa-coefficient9" |

So, these is the parameters that corresponds to each of the old CRT getters:

`RSA_get0_p()` => OSSL_PKEY_PARAM_RSA_FACTOR1
`RSA_get0_q()` => OSSL_PKEY_PARAM_RSA_FACTOR2
`RSA_get0_dmp1()` => OSSL_PKEY_PARAM_RSA_EXPONENT1
`RSA_get0_dmq1()` => OSSL_PKEY_PARAM_RSA_EXPONENT2
`RSA_get0_iqmp()` => OSSL_PKEY_PARAM_RSA_COEFFICIENT1

**Reply by nagu1137** [reply]
Thank you for your help.

Is there way we can know these kind of mappings?

**Reply by levitte** [reply]
Actually yeah, you can!  There's the manual page [EVP_PKEY-RSA(7)](https://www.openssl.org/docs/man3.2/man7/EVP_PKEY-RSA.html)

It doesn't inform you in terms of older function calls, but it does tell you what known RSA specification names the parameters correspond to.
